scylla/src/transport/topology.rs

```diff
@@ -42,6 +42,8 @@ pub(crate) struct MetadataReader {
 pub struct Metadata {
     pub peers: Vec<Peer>,
     pub keyspaces: HashMap<String, Keyspace>,
+    // This is not really part of metadata, but is required for their further interpretation.
+    pub control_connection_address: SocketAddr,
```

Comment:
The more I think about it, the more it makes sense to me to put separate translated/untranslated addresses into the `Peer` struct. With the current implementation in the PR, interpreting the address in a Peer always requires the parent Metadata to be known (or at least the control connection address from the moment the metadata was created), which is a bit cumbersome.

Keeping both translated/untranslated addresses in the Peer would translate to an easier API. You could pass an `Option<&dyn AddressTranslator>` as an argument to `query_peers` and translate the address when creating a `Peer`. A positive side effect of that is that you will only have to translate the address once, when fetching - not twice (`ClusterData::new` + `MetadataReader::update_known_peers`), and you don't need to keep the control_connection_address used during the metadata fetch.

I suggest adding an `untranslated_address: Option<IpAddr>` to the peer. The existing `address` would represent the translated address and would work as it works now in the PR. The new `untranslated_address` would always be equal to the address in the `system.peers` for non-control-connection nodes; for the control connection node, when address translation is disabled we can use the address from `system.local`, and `None` otherwise (as the address we used to connect is most likely different and we just don't know).

Finally, there is a chicken-and-egg problem here as the translator requires a `Peer` to translate the address... Let's introduce:

```rust
#[non_exhaustive] // <- so that we can add more fields in a backwards-compatible way
pub struct UntranslatedPeer {
    pub untranslated_address: SocketAddr,
}
```

and use it as an argument instead.

Reply:
At first glance, your idea seems to work very well. I'll try to implement it.

Reply:
> I suggest adding an untranslated_address: Option<IpAddr> to the peer.
Did you mean `Option<SocketAddr>` ?

Reply:
> Did you mean `Option<SocketAddr>` ?

Yes, my mistake. Edited the comment.